crates/node/src/evm.rs

```diff
@@ -252,51 +252,43 @@ impl ConfigureEvm for OdysseyEvmConfig {
-        reth_revm::primitives::GRANITE
-    } else if chain_spec.fork(OptimismHardfork::Fjord).active_at_head(block) {
-        reth_revm::primitives::FJORD
-    } else if chain_spec.fork(OptimismHardfork::Ecotone).active_at_head(block) {
-        reth_revm::primitives::ECOTONE
-    } else if chain_spec.fork(OptimismHardfork::Canyon).active_at_head(block) {
-        reth_revm::primitives::CANYON
-    } else if chain_spec.fork(OptimismHardfork::Regolith).active_at_head(block) {
-        reth_revm::primitives::REGOLITH
-    } else if chain_spec.fork(OptimismHardfork::Bedrock).active_at_head(block) {
-        reth_revm::primitives::BEDROCK
-    } else if chain_spec.fork(EthereumHardfork::Prague).active_at_head(block) {
-        reth_revm::primitives::PRAGUE
-    } else if chain_spec.fork(EthereumHardfork::Cancun).active_at_head(block) {
-        reth_revm::primitives::CANCUN
-    } else if chain_spec.fork(EthereumHardfork::Shanghai).active_at_head(block) {
-        reth_revm::primitives::SHANGHAI
-    } else if chain_spec.fork(EthereumHardfork::Paris).active_at_head(block) {
-        reth_revm::primitives::MERGE
-    } else if chain_spec.fork(EthereumHardfork::London).active_at_head(block) {
-        reth_revm::primitives::LONDON
-    } else if chain_spec.fork(EthereumHardfork::Berlin).active_at_head(block) {
-        reth_revm::primitives::BERLIN
-    } else if chain_spec.fork(EthereumHardfork::Istanbul).active_at_head(block) {
-        reth_revm::primitives::ISTANBUL
-    } else if chain_spec.fork(EthereumHardfork::Petersburg).active_at_head(block) {
-        reth_revm::primitives::PETERSBURG
-    } else if chain_spec.fork(EthereumHardfork::Byzantium).active_at_head(block) {
-        reth_revm::primitives::BYZANTIUM
-    } else if chain_spec.fork(EthereumHardfork::SpuriousDragon).active_at_head(block) {
-        reth_revm::primitives::SPURIOUS_DRAGON
-    } else if chain_spec.fork(EthereumHardfork::Tangerine).active_at_head(block) {
-        reth_revm::primitives::TANGERINE
-    } else if chain_spec.fork(EthereumHardfork::Homestead).active_at_head(block) {
-        reth_revm::primitives::HOMESTEAD
-    } else if chain_spec.fork(EthereumHardfork::Frontier).active_at_head(block) {
-        reth_revm::primitives::FRONTIER
-    } else {
-        panic!(
-            "invalid hardfork chainspec: expected at least one hardfork, got {:?}",
-            chain_spec.hardforks
+    chain_spec
+        .forks_iter()
+        .find(|(_, condition)| condition.active_at_head(block))
+        .map_or_else(
+            || {
+                panic!(
+                    "invalid hardfork chainspec: expected at least one active hardfork, got {:?}",
+                    chain_spec.hardforks
+                )
+            },
+            |(fork, _)| map_hardfork_to_spec_id(fork)
         )
+}
+
+
+/// Map hardfork to the corresponding `SpecId`.
+fn map_hardfork_to_spec_id(fork: &EthereumHardfork) -> reth_revm::primitives::SpecId {
```

Comment:
some of these are `OptimismHardfork`, which means we'd need dynamic dispatch. not really sure this is the way to go

Reply:
Given the issue with handling both EthereumHardfork and OptimismHardfork, would you recommend simplifying revm_spec by removing map_hardfork_to_spec_id altogether and directly matching on each fork type? Here’s what that would look like:

```rust
fn revm_spec(chain_spec: &ChainSpec, block: &Head) -> SpecId {
    chain_spec
        .forks_iter()
        .find(|(_, condition)| condition.active_at_head(block))
        .map_or_else(
            || {
                panic!(
                    "Invalid chainspec: expected at least one active hardfork, got {:?}",
                    chain_spec.hardforks
                );
            },
            |(fork, _)| {
                if let Some(eth_fork) = fork.downcast_ref::<EthereumHardfork>() {
                    match eth_fork {
                        EthereumHardfork::Prague => SpecId::OSAKA,
                        // (other EthereumHardfork mappings)
                    }
                } else if let Some(opt_fork) = fork.downcast_ref::<OptimismHardfork>() {
                    match opt_fork {
                        OptimismHardfork::Granite => SpecId::GRANITE,
                        // (other OptimismHardfork mappings)
                    }
                } else {
                    panic!("Unsupported hardfork: {:?}", fork);
                }
            }
        )
}
```

This approach would avoid dynamic dispatch and handle both fork types without additional casting. Do you think this is a better solution?